quail/src/main/java/com/mineinjava/quail/pathing/PathFollower.java

```diff
@@ -46,31 +54,42 @@ public void setPath(com.mineinjava.quail.odometry.path path) {
      * This does return a field-centric movement vector. It does not limit acceleration (yet)
      * @return the next movement to follow the path
      */
-    public robotMovement calculateNextDriveMovement() {
+    public RobotMovement calculateNextDriveMovement() {
         // calculate the next movement to follow the path
         Pose2d currentPose = this.localizer.getPoseEstimate();
-        double deltaAngle = util.deltaAngle(currentPose.heading, this.path.finalHeading);
+        double deltaAngle = Util.deltaAngle(currentPose.heading, this.path.getCurrentPoint().heading); // this may or may not work
+
         double turnSpeed = turnController.getOutput(0, deltaAngle);
         turnSpeed /= this.path.length();
-        turnSpeed = util.clamp(turnSpeed, -this.maxTurnSpeed, this.maxTurnSpeed);
-        Vec2d movementVector = this.path.vectorToNearestPoint(currentPose.x, currentPose.y, this.path.currentPoint);
-        movementVector.scale(this.speed/movementVector.getLength());
+        turnSpeed = Util.clamp(turnSpeed, -this.maxTurnSpeed, this.maxTurnSpeed);
+
+        Vec2d movementVector = this.path.vectorToNearestPoint(this.localizer.getPoseEstimate(), this.path.currentPoint);
+        movementVector.scale(this.speed / movementVector.getLength());
         movementVector.rotate(-currentPose.heading, false);
+
         if (movementVector.getLength() < this.precision) {
             this.path.currentPoint++;
         }
         if (this.isFinished()) {
-            return new robotMovement(0, new Vec2d(0, 0)); // the path is over
+            return new RobotMovement(0, new Vec2d(0, 0)); // the path is over
         }
-        return new robotMovement(turnSpeed, movementVector);
+
+        return new RobotMovement(turnSpeed, movementVector);
     }
+    
     /** returns true if the robot is finished following the path.
      */
     public Boolean isFinished() {
-       Pose2d currentPose = this.localizer.getPoseEstimate();
-       if (this.path.currentPoint >= this.path.points.size() - 1) {
+        if( this.localizer == null ) {
+            throw new NullPointerException("localizer is null, ensure that you have instantiated the localizer object");
+        }
```

Comment:
I feel like this should be thrown earlier before the user checks if the path is finished.